File: week#5/task#2/abstract_Animal.java
// The following code applies Abstraction: Abstract Classes
// main public class
public abstract class abstract_Animal {
    // abstract method: mandatory to be initilized in derived classes
    abstract void speak();

    // non-abstract method
    void sleep() {
        System.out.println("Sleeping...................");
    }


    public static void main(String[] args) {
        Dog dog = new Dog();
        dog.speak();
        
    }

}

class Dog extends abstract_Animal{

    void speak() {
        System.out.println("Bow bow.");
    }

    public static void main(String[] args) {
    }

}

class Cat extends abstract_Animal{
    void speak() {
        System.out.println("Meow meow.");
    }

}

File: week#5/task#2/abstract_Animal.cpp
// C++ uses abstract functions using pure virtual methods as it does not have `abstract` keyword.
// to simulate interface: you make a class where you only use abstract functions.
#include<iostream>
using namespace std;

class Animal {
// this is a pure virtual method
    virtual void speak() = 0;
    
};

class Dog : public Animal {
public:
    void speak() {
        cout<<"Bow bow bow"<<endl;
    }
};

class Cat : public Animal {
public:
    void speak() {
        cout<<"Meow meow"<<endl;
    }
};


int main() {
    Dog dog;
    dog.speak();

}
